Login

Restioson opened the pull request:
Checklist:
- [x] Basic login
- [x] Editing username, display name and password
  - [x] Implement on client
- [x] Authentication tokens
  - [x] Check token is valid before allowing use
  - [x] Clean up expired tokens periodically
  - [x] "Save password" feature on client via saving tokens
  - [x] Allow token to expire if not used for a certain duration
  - [x] Token becomes stale if not used for a certain duration and must be refreshed manually with password
  - [x] Check permissions of token before allowing actions
    - [x] Test
- [x] Extra user data (e.g profile picture)
  - [x] GDPR/POPIA/etc. considerations
- [x] Normalize and lowercase username
- [x] Compromise security (e.g what to do after compromise) 
  - [x] Check user is not locked/compromised before allowing login
  - [x] Test
- [x] Encrypt traffic between client and server
  - [x] Integrate into Docker
- [ ] Audit (cc @xacrimon)
  - [x] Preliminary
  - [ ] Final (before merge)

Restioson committed:
Docker

Restioson committed:
Asynchronise somewhat

Preparing for async db connecting

Restioson committed:
Login & database

Restioson committed:
Move hashing onto threadpool

Restioson committed:
Docker & get rid of some boxes

Restioson committed:
Bit of cleanup

gegy1000 commented on client/gtk/src/main.rs:
```diff
@@ -59,18 +59,22 @@ impl Update for Win {
     fn model(relm: &Relm<Win>, args: VertexArgs) -> VertexModel {
+        println!(
+            "Connecting to {}",
+            args.ip.clone().unwrap_or("127.0.0.1:8080".to_string())
```
Should extract this to avoid the duplication (and cloning)

Restioson replied:
Wdm?

gegy1000 replied:
`unwrap_or` is done on the ip twice. Could be extracted into its own variable

gegy1000 commented on server/src/database.rs:
```diff
@@ -0,0 +1,162 @@
+            .and_then(|mut conn| {
+                conn.client
+                    .prepare(
+                        "CREATE TABLE IF NOT EXISTS users (
```
Maybe should make these their own files and `include_str!`

Restioson replied:
Maybe indeed

Restioson replied:
May also be clearer inline

Restioson committed:
Max password length

Restioson commented:
Do we need to mitigate homoglyphs in usernames?

Restioson committed:
Change password and username

Restioson committed:
Login tokens table

Restioson committed:
Token-based auth - BROKEN CLIENT

Don't know how to do token storage on client -- will do this later

Restioson committed:
Token-based auth fix

A few things to do still - display names on client, changing names etc, and storing tokens

Restioson committed:
Max len for username and display name

Restioson committed:
Remove some unused error types

Restioson committed:
Always process username before use (comparison, storage, etc)

Restioson committed:
Use display name on client

Can't set it at user creation yet

Restioson committed:
Min password length

Restioson commented:
What extra user data do we need to store? @gegy1000

Restioson commented:
And I wonder what the implications of the GDPR and POPIA (the South African equivalent act) are for data storage, and what we need to therefore implement.

Restioson committed:
Better (more resilient) config

Restioson commented:
So far I haven't implemented any high level tools for post-compromise situations, but I have added a `compromised` boolean and a version of password hash/scheme in the database for users and login tokens.

Restioson committed:
Profile picture hosting

Restioson committed:
Require old password to change user's password

Restioson committed:
Sessions -> device (1 session per device) & revoke auth token

Restioson committed:
Revoke token

Restioson commented:
I've committed a bit more and am ticking off features according to local branch. Unfortunately I can't push yet because my Internet is down.

Restioson committed:
More stuff

Restioson commented:
I don't think so far we have incurred any special GDPR/POPIA requirements. Though, when we implement admin tools, then we will need to have some kind of way to delete users and their data.

With regards to extra user data, I've just added the capability for the server to serve profile pictures at the route `/images/profile_pictures/{id}`, served from within the `./files` directory. Later on we can flesh this out more.

Restioson committed:
Sweep tokens for expired ones

Restioson committed:
Save token

gegy1000 commented on client/backend/src/lib.rs:
```diff
@@ -1,3 +1,6 @@
+// TODO save password feature through saving tokens
```
Can remove this TODO now

gegy1000 commented on client/backend/src/lib.rs:
```diff
@@ -120,9 +134,18 @@ impl Vertex {
+    pub fn create_user(
+        &mut self,
+        username: &str,
+        display_name: &str,
```
I guess we should consider whether we should do global display name or rather have display names per server like discord?

Restioson replied:
We can have both!

Restioson committed:
Fix docker

Restioson committed:
Add /forgettoken (testing)

Restioson committed:
Some edits to config

Restioson committed:
Token and devices expires after set duration of disuse

gegy1000 commented on server/src/auth.rs:
```diff
@@ -0,0 +1,119 @@
+
+pub struct TooShort;
+
+pub fn process_username(username: &str, _config: &Config) -> String {
```
`normalize_username`?

Restioson replied:
Normalization is a unicode thing. Basically it makes the string into a consistent and comparable form. We use the most destructive form of normalization, where an anarchist circle-A will get converted to an A. The reason that we do this is to allow logins from different computers with perhaps different input languages (see what happened to Spotify), and also to make it more difficult to impersonate others using homoglyph attacks (e.g cyrillic a vs latin a).

gegy1000 replied:
I was proposing to rename it. :P 
But thanks for the explanation! I understand it better now :slightly_smiling_face:

Restioson replied:
Ah, I see. `process_username` also makes it lowercase, so it isn't just normalisation, though.

gegy1000 commented on server/src/client/server.rs:
```diff
@@ -1,100 +1,110 @@
+        // TODO could probably be optimised
+        for (user_id, device_id) in logout.list {
+            if let Some(online_devices) = self.online_devices.get(&user_id) {
+                let device_id = online_devices.iter().find(|id| **id == device_id);
```
I think this is just a `.contains`?

Restioson replied:
No, this returns the device id. Contains only checks if it exists.

gegy1000 replied:
You already have the device id

Restioson replied:
Oh lol

gegy1000 commented on server/src/client/session.rs:
```diff
@@ -208,3 +104,663 @@ impl Handler<SendMessage<ServerMessage>> for ClientWsSession {
+        user_id: UserId,
+        password: String,
+    ) -> impl Future<Item = Result<(), ServerError>, Error = MailboxError> {
+        self.database_server
```
Interaction with the database server could maybe be extracted to a function for requests specifically. Like it'd map the database errors to internal server error.

Restioson replied:
Huh? What do you mean?

gegy1000 replied:
There's a lot of repetition in these similar functions that map the error from database to an internal server error type -- I'm proposing a function that would make a request to the database via send but map the error too such that it doesn't have to be repeated

gegy1000 commented on server/src/config.rs:
```diff
@@ -0,0 +1,137 @@
+}
+
+fn max_password_len() -> u16 {
+    1000
```
not a review but hells yeah

Restioson replied:
This is just a limit to stop people from being able to DOS us by sending too many characters and expecting us to hash it, lol. All passwords are the same size after hashing.

gegy1000 replied:
very epic

gegy1000 commented on server/src/config.rs:
```diff
@@ -0,0 +1,137 @@
+    7 // 1 week
+}
+
+fn token_expiry_days() -> u16 {
```
how are these functions used?

Restioson committed:
Allow for refreshing of tokens

Fixed bug that `last_used` would never be updated

Restioson committed:
Tokens become stale after disuse for a certain period of time and need to be refreshed

Also capped token length at 45 char to avoid DOS (client could previously send a large token and force the server to hash it all)

Restioson committed:
Check whether client has correct permissions before performing action

Restioson committed:
Review comments

Restioson committed:
Some things

- device names on tokens
- token len > max to token len < max (oops!!)
- fixed review comment about duplicated error handling

Restioson committed:
Change 1 print

Restioson commented:
@gegy1000 any further review comments?

Restioson committed:
TLS between client and server

Still need to integrate into Docker

gegy1000 commented:
Nope, all looks great to me! 🙂

Restioson committed:
Fix permissions check

Restioson committed:
Some updates to dependencies

Restioson committed:
Docker works now + some edits

Restioson commented:
Since @xacrimon is busy, perhaps we should merge this into a development branch and then wait for a review to merge that into master?

Gegy commented:
Sounds good 👍
